File: 剑指Offer/13.机器人的运动范围.py
class Solution:
    def __init__(self):
        self.addX = [1, -1, 0, 0]
        self.addY = [0, 0, 1, -1]
        self.cellsCounts = 0
    
    def getDigitSum(self, x: int, y: int) -> int:
        res = 0
        while x != 0 or y != 0:
            if x:
                res += x % 10
                x //= 10
            if y:
                res += y % 10
                y //= 10
        return res
            
    def backtrack(self, visited: List[List[int]], m: int, n: int, k: int, x: int, y: int) -> None:
        for i in range(4):
            newX = x + self.addX[i]
            newY = y + self.addY[i]
            if newX >= 0 and newX < m and newY >= 0 and newY < n and visited[newX][newY] == 0:
                if self.getDigitSum(newX, newY) <= k:
                    self.cellsCounts += 1
                    visited[newX][newY] = 1
                    self.backtrack(visited, m, n, k, newX, newY);

    def movingCount(self, m: int, n: int, k: int) -> int:
        visited = [[0 for i in range(n)] for j in range(m)]
        visited[0][0] = 1
        self.cellsCounts += 1
        self.backtrack(visited, m, n, k, 0, 0)
        return self.cellsCounts

File: 剑指Offer/13.机器人的运动范围.cpp
class Solution {
    int cells = 0;
    int addX[4] = {1, -1, 0, 0};
    int addY[4] = {0, 0, 1, -1};
    int getDigitalSum(int x, int y){
        int res = 0;
        while (x || y) {
            if (x){
                res += x % 10;
                x /= 10;
            }
            if (y) {
                res += y % 10;
                y /= 10;
            }
        }
        return res;
    }

    void backtrack(vector<vector<int>>& visited, int m, int n, int k, int x, int y) {
        for (int i = 0; i < 4; ++i) {
            int newX = x + addX[i], newY = y + addY[i];
            if (newX >= 0 && newX < m && newY >= 0 && newY < n && visited[newX][newY] == 0) {
                if (getDigitalSum(newX, newY) <= k) {
                    cells++;
                    visited[newX][newY] = 1;
                    backtrack(visited, m, n, k, newX, newY);
                }
            }
        }
    }

public:
    int movingCount(int m, int n, int k) {
        vector<vector<int>> visited(m, vector<int>(n, 0));
        visited[0][0] = 1;
        cells++;
        backtrack(visited, m, n, k, 0, 0);
        return cells;
    }
};